File: token-validation-core/src/main/java/no/nav/security/token/support/core/validation/DefaultJwtClaimsVerifier.java
package no.nav.security.token.support.core.validation;

import com.nimbusds.jose.proc.SecurityContext;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.proc.BadJWTException;
import com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier;
import com.nimbusds.jwt.util.DateUtils;
import com.nimbusds.openid.connect.sdk.validators.BadJWTExceptions;

import java.util.Date;
import java.util.Set;

/**
 * Extends {@link com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier} with a time check for the issued at ("iat") claim.
 * The claim is only checked if it exists in the given claim set.
 */
public class DefaultJwtClaimsVerifier<C extends SecurityContext> extends DefaultJWTClaimsVerifier<C> {

    public DefaultJwtClaimsVerifier(final Set<String> acceptedAudience,
                                    final JWTClaimsSet exactMatchClaims,
                                    final Set<String> requiredClaims,
                                    final Set<String> prohibitedClaims) {
        super(acceptedAudience, exactMatchClaims, requiredClaims, prohibitedClaims);
    }

    @Override
    public void verify(final JWTClaimsSet claimsSet, final C context) throws BadJWTException {
        super.verify(claimsSet, context);

        Date iat = claimsSet.getIssueTime();
        if (iat != null) {
            Date now = new Date();
            if (!iat.equals(now) && !DateUtils.isBefore(iat, now, super.getMaxClockSkew())) {
                throw BadJWTExceptions.IAT_CLAIM_AHEAD_EXCEPTION;
            }
        }
    }
}

File: token-validation-core/src/main/kotlin/no/nav/security/token/support/core/validation/DefaultJwtClaimsVerifier.kt
package no.nav.security.token.support.core.validation

import com.nimbusds.jose.proc.SecurityContext
import com.nimbusds.jwt.JWTClaimsSet
import com.nimbusds.jwt.proc.BadJWTException
import com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier
import com.nimbusds.jwt.util.DateUtils
import com.nimbusds.openid.connect.sdk.validators.BadJWTExceptions.IAT_CLAIM_AHEAD_EXCEPTION
import java.util.Date

/**
 * Extends [com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier] with a time check for the issued at ("iat") claim.
 * The claim is only checked if it exists in the given claim set.
 */
class DefaultJwtClaimsVerifier<C : SecurityContext>(acceptedAudience : Set<String?>?, exactMatchClaims : JWTClaimsSet, requiredClaims : Set<String>, prohibitedClaims : Set<String>) : DefaultJWTClaimsVerifier<C>(acceptedAudience, exactMatchClaims, requiredClaims, prohibitedClaims) {

    @Throws(BadJWTException::class)
    override fun verify(claimsSet: JWTClaimsSet, context: C?) {
        super.verify(claimsSet, context)
        claimsSet.issueTime?.let { iat ->
            if (!DateUtils.isBefore(iat, Date(), maxClockSkew.toLong())) {
                throw IAT_CLAIM_AHEAD_EXCEPTION
            }
        }
    }
}